test_app/.browserslistrc

```diff
@@ -0,0 +1 @@
+defaults
```

Comment:
What is this file about?

Reply:
This is a separated Browserslist config file for targeting the browsers that the program support. The defaults query specified in this file will give the users the most reasonable configuration (> 0.5%, last 2 versions, Firefox ESR, not dead). The file is required to be put under the parent or current directory.

Reference: https://github.com/browserslist/browserslist#readme

Reply:
1. You say: `will give the users the most reasonable configuration`. Define `reasonable`. Why do you think these level of browsers are `reasonable`?
2. What are all the browser & its version that are supported? What is the command to get it?
3. How does having this tie to the rails app supporting the said versions only & not others?

Reply:
1. The config for defaults is (> 0.5%, last 2 versions, Firefox ESR, not dead)
**> 0.5%**
- select the most popular versions ( top 0.5 % ) of the browser 
**last 2 versions:** 
- select the last two versions of the browser
**Firefox ESR:** 
- the latest Firefox Extended Support Release.
**not dead:** 
- remove browser versions without official support in the last 24 months. 

This config ensures the browser will include the not dead old popular versions, while supporting the newer release version.

2. Command: npx browserslist
and_chr 96
and_ff 95   
and_qq 10.4 
and_uc 12.12
android 96  
baidu 7.12
chrome 96
chrome 95
chrome 94
edge 96
edge 95
firefox 95
firefox 94
firefox 91
firefox 78
ie 11
ios_saf 15.2
ios_saf 15.0-15.1
ios_saf 14.5-14.8
ios_saf 14.0-14.4
ios_saf 12.2-12.5
kaios 2.5
op_mini all
op_mob 64
opera 82
opera 81
safari 15.2
safari 15.1
safari 14.1
safari 13.1
samsung 15.0
samsung 14.0

3.  The transpilers or bundlers will categorize the browsers into different categories and generate seperate bundles for different browsers. For example:

- Legacy Bundle: Contains polyfills (  code that implements a feature on web browsers that do not support the feature.) , larger bundle size, compatible with old browsers without ES6 support.
- Modern Bundle: Smaller bundle size, optimized for modern browsers.

Our entry point (e.g index.html) is generated in a way that it'll load the required bundles according to the current browser being used by a user.  In the above two groups, the legacy bundle will be loaded when users are using old browsers that don't have ES6 support whereas the modern bundle will be loaded if users are using modern web browsers that have the recent javascript ES6 support.

Reply:
1. Good
2. Good
3. You didn't explain how does it happen in rails.

Reply:
In rails, we have webpacker gem to compile all the assets and generate static files for production. When compiling the javascript codes, browserlist will tell webpacker to compile according to the browser version in browserlistrc config and prevent webpacker to compile it to old syntax

Reply:
How does webpacker does the compiling? What do you mean by `compile to old syntax`?

Reply:
The webpacker will compile any JavaScript file in the app/javascript/packs directory ( default entry path ) to its own pack file. And in our frontend html.erb file, we can include tag <%= javascript_pack_tag "application" %> to import our application pack ( If you have a application.js file inside the packs folder, then webpacker will generate a application pack for you ). To require any script, we can import it in the application.js ( the file from above example ). We can have javascript_pack_tag for javascript, stylesheet_pack_tag for stylesheet ( css... ), image_pack_tag for images and asset_pack_tag for our assets.

The old syntax here I refer to older ECMAScript standards like ES5. In webpacker, the webpack uses babel to transpile the javascript for ES6 to ES5, so the javascript can be used by the older version browser.

Reply:
1. You say `the webpacker`. How? Takkan magic? Sure got some logical place that this is defined. 
2. `app/javascript/packs directory ( default entry path )`: Where is this defined?
3. What does `transpile` means?

Reply:
### You say the webpacker. How? Takkan magic? Sure got some logical place that this is defined.
### app/javascript/packs directory ( default entry path ): Where is this defined?

Webpacker will be automatically installed in the recent version of Rails and will be automatically invoked for node versions that higher than 10.17.0. It is a tool that uses webpack to manage app-like JavaScript. In the recent version of rails, Webpacker compiles automatically on-demand in development when a Rails page loads, which means you no need to run a Webpacker process separately. 

Every entry points should be placed in the app/javascript/packs directory and the other files ( css, static asstes and your javascript files,  etc ) should be placed outside the packs directory but stored inside the app/javascript directory. It was recommended to structure the files as below as example shown in the webpacker github: https://github.com/rails/webpacker

```
app/javascript:
  ├── packs:
  │   # only webpack entry files here
  │   └── application.js
  │   └── application.css
  └── src:
  │   └── my_component.js
  └── stylesheets:
  │   └── my_styles.css
  └── images:
      └── logo.svg

```

Any JavaScript file in the app/javascript/packs directory ( default entry path ) will be compiled by webpacker to its own pack file. We can include tag <%= javascript_pack_tag "application" %> in our html.erb file's header section to import our application pack. The default pack created for you by Webpacker will link to Rails' default JavaScript packages if they have been included in the project.

To change the default path for input or output, we can adjust the source_path (default app/javascript) and source_entry_path (default packs) in the **config/webpacker.yml file**. Below is a snippet of webpacker.yml file. 

```
default: &default
  source_path: app/javascript
  source_entry_path: packs
  ...

development:
 <<: *default
  ...

test:
  <<: *default
```
The webpacker.yml includes config for different environment such as development and test, their config was inherited from the default. 

### What does transpile means?
ECMA Script is the JavaScript language specification and the ECMA Script 2015 ES6 version is the stable version which works fine in most of browsers. However, ES6 came with some new features (promises, arrow functions, etc) which might not be supported by some browsers that do not support ES6.  ( same for ES7, ES8 and following versions ).

Therefore, to work with old browsers, there is a need for us to convert the javascript to ES5 and polyfill for the old browsers that does not support recent versions of javascript ( ES6 onwards ). In the case of the rails, the webpacker will use Babel which is called a transpiler to transpile the code in the ECMA Script version that the browsers support. We can now write the javascript in the latest standard with it.

Reply:
Still didn't explain how rails work with webpacker. All you mentioned are the configuration that is set / mentioned in the docs on where to put your file so it magically work.

What I am asking is how does the magic work.

Reply:
### How does webpacker run?
When installing rails/webpacker, rails/webpacker comes along with two bin ( binstub ) files called webpack ( production ) and webpack-dev-server ( development ) in the bin folder. Inside the bin files, it is going to call WebpackRunner with the ARGV parameter that passed into it.  
```
Dir.chdir(APP_ROOT) do
  Webpacker::DevServerRunner.run(ARGV)
end
```

When the runner gets called, the initialize function that is defined in the https://github.com/rails/webpacker/blob/master/lib/webpacker/runner.rb. will be executed. One thing to take note of here is the webpacker will choose the appropriate js file for the configuration of the webpack according to the NODE_ENV that is assigned into @webpacker_config. If the NODE_ENV is development, then the config\webpack\development.js will be chosen for setting up the environment. There are other js files such as production js specialize for production, test.js specialize for testing and there is another environment.js which served as a common environment config that required by other js files, we can also customize that file to apply to all other different environments. The environment config is required at the last second line of the other environment js files at ```const environment = require('./environment')```. In the production.js, test.js, and development.js, the last line of them is the ```module.exports = environment.toWebpackConfig() ``` which will generate a webpack config that is defined in https://webpack.js.org/configuration/ and generates the manifest.json. This configuration is then processed by webpack based upon its defined properties and generates the manifest.json, compiled JSS, CSS, other static assets, etc .

### How webpacker compile assets ( production ):
For production, we must run the following command:
```bundle exec rails assets:precompile```
In production, Webpacker will add task ```webpacker:compile``` to task ```assets:precompile``` .  The detailed procedure can be inspected with command ```rails assets:precompile –trace``` :

```
** Invoke assets:precompile (first_time)
** Invoke assets:environment (first_time)
** Execute assets:environment
** Invoke environment (first_time)       
** Execute environment
** Invoke yarn:install (first_time)
yarn install v1.22.17
[1/4] Resolving packages...
….
** Execute assets:precompile
** Invoke webpacker:compile (first_time)
** Invoke webpacker:verify_install (first_time)
** Invoke webpacker:check_node (first_time)
** Execute webpacker:check_node
** Invoke webpacker:check_yarn (first_time)
** Execute webpacker:check_yarn
** Invoke webpacker:check_binstubs (first_time)
** Execute webpacker:check_binstubs
** Execute webpacker:verify_install
** Invoke environment
** Execute webpacker:compile
```

From the trace, it can be seen that the bundle exec rails assets:precompile is run before webpacker:compile (yarn install -> Execute assets:precompile -> Invoke webpacker:compile”, after yarn install v1.22.17, webpack compile is running) so if building a pipeline by running assets:precompile , it compiles the assets first then compiles the files specified by the webpack. Because the webpack package is part of package.json , yarn will automatically install the webpack so it can compile the JS code.

### How webpacker compile assets ( development ):
For development, we have several options but the most recommended one is to use the webpack-dev-server. This webpack-dev-server process can be created by using the ``` ruby .\bin\webpack-dev-server```,  the port it is running on and other configurations are configured in the webpacker.yml. This is better because when a web request arrives, Rails will skip calling webpack compilation since it will detect the webpack-dev-server is running to save the time for compilation. Requests for webpack assets will proxy to the webpack-dev-server. 
The server proxy is set up in the webpacker/lib/webpacker/railtie.rb
```
  initializer "webpacker.proxy" do |app|
    if (Webpacker.config.dev_server.present? rescue nil)
      app.middleware.insert_before 0,
        Rails::VERSION::MAJOR >= 5 ?
          Webpacker::DevServerProxy : "Webpacker::DevServerProxy", ssl_verify_none: true
    end
end
```
### How does view helper work?
The webpacker will create the webpack config object and run the webpack. This will create this manifest.json, JSS, CSS, Fonts, Images, etc the static assets, that are going to go into Rails. It is going to create the HTML sent over to Rails. The view helpers defined in ( https://github.com/rails/webpacker/blob/master/lib/webpacker/helper.rb ) will generate the HTML tag for the HTML front end also Rails is going to be ready to serve those statically created assets. The tags can take reference to the actual assets through manifest.json file inside the public folder as default config in webpacker.yml file. A snippet of manifest.json is shown below:
Manifest.json:
```
{
  "application.js": "/packs/js/application-47eead385c4514ca149d.js",
  "application.js.map": "/packs/js/application-47eead385c4514ca149d.js.map",
  "entrypoints": {
    "application": {
      "js": [
        "/packs/js/application-47eead385c4514ca149d.js"
      ],
      "js.map": [
        "/packs/js/application-47eead385c4514ca149d.js.map"
      ]
   }
…
}
```
In the manifest.json, it has a property called entrypoints which includes the path for packs that was created out of the files placed in the app\javascript\packs.

Reply:
> One thing to take note of here is the webpacker will choose the appropriate js file for the configuration of the webpack according to the NODE_ENV that is assigned into @webpacker_config in the initialize function

There is no where that you set `NODE_ENV` specifically. How does rails set the ENV for webpacker to use?

2. What is a `server proxy` ?

> `"/packs/js/application-47eead385c4514ca149d.js"`
Why is the file name `application-<?>` ?
And what is the `<?>` ?

Reply:
### How does rails set the ENV for webpacker to use?
NODE_ENV can either be set to production, development or test. We don't need to create additional environment files inside config/webpacker/* and instead use webpacker.yml to load different configurations using RAILS_ENV.
Below is a example to select development as Node environment
`RAILS_ENV=development NODE_ENV=development bundle exec rails assets:precompile
`
### What is a server proxy ?
Server proxy is a gateway between the user and the website, it encrypts and makes changes to the web request made by users and forwards the data from the website back to the users. It also helps filter and protects you from browsing a malicious website.

### Why is the file name application-<?> ?
### And what is the <?> ?

Webpack provides a method of templating the filenames using bracketed strings called substitutions. It will add a unique hash based on the content of an asset. And that hash is what the ? of <?>  is. And the hash is put after the name of the pack the webpack generated. 

To understand why we need the hash, we have to understand the problem first. When using webpack to bundle our assets, the output will be generated inside public/packs and once the contents of /public/packs have been deployed to a server, clients (typically browsers) will hit that server to grab the site and its assets. And the last step is time-consuming. With caching, we can cache the unchanged files on the client-side and reduce unnecessary traffic.  Only the files that you made changes to will get a new hash and refetch from server, the other files that have no changes will remain their hash and cache on the client-side.